File: backend/src/main/scala/domains/blog/objects/response/BlogDetail.scala
package domains.blog.objects.response

import domains.blog.objects.*

import domains.user.objects.UserIdentity
import io.circe.{Decoder, Encoder}
import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}
import shared.objects.access.ResourceVisibilityPolicy

import java.time.Instant
import scala.util.Try

/** 博客详情响应，包含正文、关联题目、投票信息和评论列表。 */
final case class BlogDetail(
  id: BlogId,
  title: BlogTitle,
  content: BlogContent,
  author: UserIdentity,
  visibilityPolicy: ResourceVisibilityPolicy,
  relatedProblems: List[BlogProblemReference],
  score: Int,
  viewerVote: Option[BlogVote],
  comments: List[BlogCommentSummary],
  createdAt: Instant,
  updatedAt: Instant
)

/** 提供博客详情 JSON codec，并显式处理 Instant 字符串格式。 */
object BlogDetail:
  private given Encoder[Instant] = Encoder.encodeString.contramap(_.toString)
  private given Decoder[Instant] = Decoder.decodeString.emap { value =>
    Try(Instant.parse(value)).toEither.left.map(_.getMessage)
  }

  given Encoder[BlogDetail] = deriveEncoder[BlogDetail]
  given Decoder[BlogDetail] = deriveDecoder[BlogDetail]

File: frontend/src/objects/blog/response/BlogDetail.ts
import type { UserIdentity } from '@/objects/user/UserIdentity'
import type { BlogCommentSummary } from '@/objects/blog/response/BlogCommentSummary'
import type { BlogContent } from '@/objects/blog/BlogContent'
import type { BlogId } from '@/objects/blog/BlogId'
import type { BlogProblemReference } from '@/objects/blog/BlogProblemReference'
import type { BlogTitle } from '@/objects/blog/BlogTitle'
import type { BlogVote } from '@/objects/blog/BlogVote'
import type { ResourceVisibilityPolicy } from '@/objects/shared/access/ResourceVisibilityPolicy'

/** 博客详情响应；包含正文、关联题目、当前用户投票和评论列表。 */
export type BlogDetail = {
  id: BlogId
  title: BlogTitle
  content: BlogContent
  author: UserIdentity
  visibilityPolicy: ResourceVisibilityPolicy
  relatedProblems: BlogProblemReference[]
  score: number
  viewerVote: BlogVote | null
  comments: BlogCommentSummary[]
  createdAt: string
  updatedAt: string
}
